src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt

```diff
@@ -226,101 +202,220 @@ class IncomingPaymentHandler(val nodeParams: NodeParams, val db: IncomingPayment
+                                    logger.warning { "rejecting on-the-fly funding: too many parts (${payment.parts.size} > ${nodeParams.maxAcceptedHtlcs}" }
+                                    nodeParams._nodeEvents.emit(LiquidityEvents.Rejected(payment.amountReceived, 0.msat, LiquidityEvents.Source.OffChainPayment, LiquidityEvents.Rejected.Reason.TooManyParts(payment.parts.size)))
+                                    rejectPayment(payment, incomingPayment, TemporaryNodeFailure)
+                                }
+                                willAddHtlcParts.isNotEmpty() -> when (val result = validateOnTheFlyFundingRate(willAddHtlcParts.map { it.amount }.sum(), remoteFeatures, currentFeeCredit, currentFeerate, remoteFundingRates)) {
+                                    is Either.Left -> {
+                                        logger.warning { "rejecting on-the-fly funding: reason=${result.value.reason}" }
+                                        nodeParams._nodeEvents.emit(result.value)
+                                        rejectPayment(payment, incomingPayment, TemporaryNodeFailure)
+                                    }
+                                    is Either.Right -> {
+                                        val (requestedAmount, fundingRate) = result.value
+                                        val addToFeeCredit = run {
+                                            val featureOk = Features.canUseFeature(nodeParams.features, remoteFeatures, Feature.FundingFeeCredit)
+                                            // We may need to use a higher feerate than the current value depending on whether this is a new channel or not,
+                                            // and whether we have enough balance. We keep adding to our fee credit until we reach the worst case scenario
+                                            // in terms of fees we need to pay, otherwise we may not have enough to actually pay the liquidity fees.
+                                            val maxFeerate = currentFeerate * AddLiquidityForIncomingPayment.SpliceWithNoBalanceFeerateRatio
+                                            val maxLiquidityFees = fundingRate.fees(maxFeerate, requestedAmount, requestedAmount, isChannelCreation = true).total.toMilliSatoshi()
+                                            val feeCreditThreshold = when (val policy = nodeParams.liquidityPolicy.value) {
+                                                LiquidityPolicy.Disable -> maxLiquidityFees
+                                                is LiquidityPolicy.Auto -> maxLiquidityFees.min(policy.maxAllowedFeeCredit)
+                                            }
+                                            val amountBelowThreshold = (willAddHtlcParts.map { it.amount }.sum() + currentFeeCredit) <= feeCreditThreshold
+                                            featureOk && amountBelowThreshold
```

Comment:
I had a hard time understanding this, due to mixing the `maxLiquidityFees` and `maxAllowedFeeCredit` which are separate concerns. How about:

```suggestion
                                            val maxLiquidityFees = fundingRate.fees(maxFeerate, requestedAmount, requestedAmount, isChannelCreation = true).total.toMilliSatoshi()
                                            val maxFeeCredit = when (val policy = nodeParams.liquidityPolicy.value) {
                                                LiquidityPolicy.Disable -> 0.msat
                                                is LiquidityPolicy.Auto -> policy.maxAllowedFeeCredit
                                            }
                                            val nextFeeCredit = willAddHtlcParts.map { it.amount }.sum() + currentFeeCredit
                                            val cannotCoverLiquidityFees = nextFeeCredit <= maxLiquidityFees
                                            val isBelowMaxFeeCredit = nextFeeCredit <= maxFeeCredit
                                            val assessment = featureOk && cannotCoverLiquidityFees && isBelowMaxFeeCredit
                                            logger.info { "fee credit assessment: result=$assessment featureOk=$featureOk cannotCoverLiquidityFees=$cannotCoverLiquidityFees isBelowMaxFeeCredit=$isBelowMaxFeeCredit (nextFeeCredit=$nextFeeCredit,maxLiquidityFees=$maxLiquidityFees,maxFeeCredit=$maxFeeCredit)" }
                                            assessment
```